### PR TITLE
[improvement]Use phmap::flat_hash_set in AggregateFunctionUniq

### PR DESCRIPTION
**Pull request by mrhhsg**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Test on clikcbench with SQL:
```sql
SELECT COUNT(DISTINCT SearchPhrase), count(distinct userid) FROM hits;
```
Execution time reduced to 7.60 sec from 9.73 sec.

## Checklist(Required)

1. Type of your changes:
    - [x] Improvement
    - [ ] Fix
    - [ ] Feature-WIP
    - [ ] Feature
    - [ ] Doc
    - [ ] Refator
    - [ ] Others: 
2. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
3. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
5. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
6. Are there any changes that cannot be rolled back:
    - [x] Yes
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/aggregate_functions/aggregate_function.h [hunk 1]
```diff
@@ -224,6 +228,9 @@ class IAggregateFunctionHelper : public IAggregateFunction {
                                                      arena);
         }
     }
+
+    virtual void deserialize_and_merge(AggregateDataPtr __restrict place, BufferReadable& buf,
+                                       Arena* arena) const override {}
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function.h:233]: Does we need add some check at here?

File: be/src/vec/aggregate_functions/aggregate_function_uniq.h [hunk 2]
```diff
@@ -111,16 +110,82 @@ class AggregateFunctionUniq final
...
+        set.rehash(size + set.size());
+
+        for (size_t i = 0; i < size; ++i) {
+            KeyType ref;
+            read_pod_binary(ref, buf);
+            set.insert(ref);
+        }
     }
 
     void deserialize(AggregateDataPtr __restrict place, BufferReadable& buf,
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_uniq.h:177]: Seems only diffrent with `deserialize_and_merge` is set.rehash, maybe we can do some abstract here?

Comment (mrhhsg) [reply be/src/vec/aggregate_functions/aggregate_function_uniq.h:177]: Yes, `deserialize` just does the same thing as `deserialize_and_merge`

**Review by kpfly — DISMISSED** [review]
LGTM

File: be/src/vec/aggregate_functions/aggregate_function_uniq.h [hunk 3]
```diff
@@ -109,18 +102,96 @@ class AggregateFunctionUniq final
...
                Arena*) const override {
-        this->data(place).set.merge(this->data(rhs).set);
+        auto& rhs_set = this->data(rhs).set;
+        if (rhs_set.size() == 0) return;
+
+        auto& set = this->data(place).set;
+        set.rehash(set.size() + rhs_set.size());
+
+        for (auto elem : rhs_set) {
+            set.insert(elem);
```
Comment (wangbo) [inline be/src/vec/aggregate_functions/aggregate_function_uniq.h:151]: phmap has merge method, is that same with insert one by one?

File: be/src/vec/aggregate_functions/aggregate_function_uniq.h [hunk 4]
```diff
@@ -109,18 +102,96 @@ class AggregateFunctionUniq final
...
+            set.insert(keys[i]);
+        }
     }
 
     void serialize(ConstAggregateDataPtr __restrict place, BufferWritable& buf) const override {
-        this->data(place).set.write(buf);
+        auto& set = this->data(place).set;
+        write_var_uint(set.size(), buf);
+        for (const auto& elem : set) {
+            write_pod_binary(elem, buf);
```
Comment (wangbo) [inline be/src/vec/aggregate_functions/aggregate_function_uniq.h:173]: How about add a Todo here; After phmap is included in BE's code. We can  serialize phmap in copy way

File: be/src/vec/aggregate_functions/aggregate_function_uniq.h (hunk 3 above)
Comment (mrhhsg) [reply be/src/vec/aggregate_functions/aggregate_function_uniq.h:151]: Yes, `phmap::merge` inserts elements one by one and it requires the src(rhs) is not constant.

**Review by wangbo — APPROVED** [review]
LGTM

**Pull request merged**